drivers/timer/nrf_rtc_timer.c

```diff
@@ -117,19 +120,39 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 	ticks = MAX(MIN(ticks - 1, (s32_t)MAX_TICKS), 0);
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	u32_t cyc, t = counter();
+	u32_t cyc, t2, t = counter();
 
 	/* Round up to next tick boundary */
-	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
+	cyc = ticks * CYC_PER_TICK + 1 + counter_sub(t, last_count);
 	cyc += (CYC_PER_TICK - 1);
 	cyc = (cyc / CYC_PER_TICK) * CYC_PER_TICK;
 	cyc += last_count;
 
-	if (counter_sub(cyc, t) < MIN_DELAY) {
-		cyc += CYC_PER_TICK;
+	/* Per NRF docs, the RTC is guaranteed to trigger a compare
+	 * event if the comparator value to be set is at least two
+	 * cycles later than the current value of the counter.  So if
+	 * we're more than two cycles out, we can set it blindly.  If
+	 * not, check the time again immediately before setting: it's
+	 * possible we "just missed it" and can flag an immediate
+	 * interrupt.  Otherwise, there's no way to get an interrupt
+	 * at the right time and we have to slip the event by one
+	 * clock cycle (or we could spin, but this is a slow clock and
+	 * spinning for a whole cycle can be thousands of
+	 * instructions!)
+	 */
+	if (counter_sub(cyc, t) > 2) {
+		set_comparator(cyc);
+	} else {
+		t2 = counter();
+		set_comparator(cyc);
+
+		if (cyc == t2) {
+			NVIC_SetPendingIRQ(RTC1_IRQn);
+		} else {
+			set_comparator(cyc + 1);
```

Comment:
But I have to agree with @pizi-nordic about this. `t` is no longer relevant; it's fine.  But what's preventing `COUNTER` from incrementing between the read of `t2` and the set of the comparator on the next line?  The fact that `cyc-t2` is at least 2 does not appear to guarantee that at the instant `COMPARE` was actually set `COUNTER` was *still* 2 below `cyc`.

Does something like this make sense?

```
t2 = counter();
set_comparator(cyc);
unsigned delta = counter_sub(cyc, t2);
if (delta == 0) {
   // Missed it but deadline has been reached, can force IRQ
} else if (delta <= 2) {
   // Maybe missed it; deadline hasn't been reached so can't force IRQ.
   // Set to next counter value guaranteed to be in time.
   set_comparator(t2 + 3);
} else {
   // At least 2 cycles out, we're fine
}
```

Reply:
IMHO we should check the current state of the counter AFTER we set compare value. If the current value is ahead of the alarm we are safe (ahead = 2 ticks or more), otherwise we either set the alarm and was pre-empted (we can check this checking IRQ pending bit) or we missed the alarm.